### ISSUE TITLE
Tensorboard Embedding wrong images/labels

### ISSUE THREAD

**Issue by lucabergamini** [lanpa/tensorboard-pytorch#12]
It's me again :)
I'm running some embedding visualization on the MNIST dataset.

My metadata.tsv 
```
0
0
0
0
0
0
0
0
1
1
1
1
1
1
1
1
2
2
2
2
2
2
2
2
3
3
3
3
3
3
3
3
4
4
4
4
4
4
4
4
5
5
5
5
5
5
5
5
6
6
6
6
6
6
6
6
7
7
7
7
7
7
7
7
8
8
8
8
8
8
8
8
9
9
9
9
9
9
9
9
```
My sprite.png
![sprite](https://user-images.githubusercontent.com/27865235/29311014-5e319100-81b0-11e7-8360-562a870bb6c7.png)

How can this possibly happen :) ?
![wrong_label](https://user-images.githubusercontent.com/27865235/29311061-97ca9e84-81b0-11e7-878b-040cdef78356.png)

I've checked the tensorboard doc and everything seems disposed as requested..

**Comment by lucabergamini** [comment]
Ok, I think I understand what is happening. according to the [tensorboard embedding doc]( https://www.tensorflow.org/get_started/embedding_viz) 

> The sprite should have the **same** number of rows and columns 

There is no such control in the code right now, and this explains why TB is link wrong images and labels (although it is a poor implementation from them to expect a squared image IMHO) . In fact if you:

- Get the images number ```n_i```
- Find the minimum squared sprite dimensions which can hold the images  using ``` (int(ceil(n_i**0.5))))**2 ``` 
- Concat black images at the end of the valid images
- Call ``` make_grid ``` using the square root size (which now does exist as integer)

Everything should works. You can see the different results for a batch of 10 images below
![sprite](https://user-images.githubusercontent.com/27865235/29320254-e253c9fa-81d6-11e7-83b5-149846fb6c19.png)

![sprite](https://user-images.githubusercontent.com/27865235/29320263-e908b558-81d6-11e7-95cb-ad1c5720a6c1.png)

The first one produces wrong association, while the second one works!

It should be easy to implement it in your code, however to make it works the line 99 
``` assert mat.size(0)==label_img.size(0), '#images should equal with #data points' ```
has to be removed, since now the dimensions are not equal

**Comment by lanpa** [comment]
great!
But I think you should fix:
https://github.com/lanpa/tensorboard-pytorch/blob/master/tensorboard/embedding.py#L15
and make `nrow` correct.
Would you like to send a PR?

**Comment by lucabergamini** [comment]
I' m working on that function indeed. Just one question:

- What is the use of  ```xx```?

**Comment by lanpa** [comment]
https://github.com/lanpa/tensorboard-pytorch/blob/master/tensorboard/embedding.py#L17

Just a dirty fix. lol

**Comment by lucabergamini** [comment]
I make a PR as soon as I learn how it works :D

**Comment by lanpa** [comment]
changing `nrow = int(math.floor(math.sqrt(label_img.size(0))))`
to `nrow = int(math.ceil(math.sqrt(label_img.size(0))))` would be OK. Am I right?

**Comment by lucabergamini** [comment]
no, beacuse you also have to concat black images to the original ones. 
Should i fork the repo and then make a pr?

**Comment by lanpa** [comment]
Yes, and please have a test on the function before PR.

**Comment by lucabergamini** [comment]
I'm getting an unexpected behaviour...the label are linked correctly, but the images of the first column (except for the first element, i.e m[0,0] ) are shown as black. Any idea?

**Comment by lucabergamini** [comment]
Fixed and PR made :)

**Comment by lanpa** [comment]
Thanks

**Comment by lanpa** [comment]
Should work now. Thanks!
